Fix extend+exclude interaction

agraul opened the pull request:
### What does this PR do?
Exclude declarations that are specified in excluded SLS files are ignored. Users don't expect that exluded SLS files have any effect at all. The PR is a bit rough around the edges, I don't mind cleaning it up a bit but first I'd like to have feedback on the approach I took.

I haven't yet checked if `None` can be in `__exclude__`. If that never happens, I will remove the sentinel object.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/62082

### Previous Behavior
An SLS file that is excluded affects other states if it contains an extend declaration.

### New Behavior
An  SLS file that is excluded does not extend other states.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Always

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

agraul committed:
Ignore extend declarations from excluded sls files

sls files that are excluded should not affect other sls files by
extending their states. Exclude statements are processed very late in
the state processing pipeline to ensure they are not overridden. By that
time, extend declarations are already processed.

Luckily, it's not necessary to change much, during the extend
declarations processing it is easy to check if the sls file that
contains a given extend declaration is excluded.

agraul committed:
Use both test sls files

agraul committed:
Test that excluded sls files can't extend others

cmcmarrow reviewed (commented):
Thank you for writing tests. For me the pull request looks really nice. I don't know if __exclude__ can hold none. I'll try and find that information for you. But It does not hurt to have it.

agraul commented:
@cmcmarrow thanks for looking into that open question. Looking at the code again, I think I can remove the sentinel object either way. When I added it, I was doing `if body.get("__sls__", sentinel_sls) in high.get("__exclude__", [])`, but then I realized that the `"__exclude__"` list can contain dictionaries with "id" or "sls" keys and added the logic to build `sls_excludes`. And `sls_excludes` can not include `None`.

Ch3LL commented on changelog/62848.fixed:
```diff
@@ -0,0 +1 @@
+Clarify pkg.installed's pkg_verify option documentation
```
I think this is the wrong changelog file and description for this change

agraul replied:
Whoops, how did that sneak in? :facepalm: I'll fix this today.
